### ISSUE TITLE
Extending rule options in derived configs?

### ISSUE THREAD

**Issue by platinumazure** [eslint/eslint#9192]
This comes up fairly frequently: Someone extends a shareable config or plugin config which might use `no-restricted-syntax` or similar to lint certain patterns. A user may wish to alter the list of linted patterns to add one more pattern; however, right now, this means copying the whole list from the shareable config or plugin config, which basically defeats the purpose of using a shareable config in the first place.

I wanted to get the ball rolling on discussing how we might address this with one simple, incomplete, likely flawed proposal. I'm hoping that we can come up with something good after a spirited discussion.

My proposal:

* Augment new-style rule support in core to allow an `extendConfiguration` method (in line with `meta` property and `create` method).
* The `extendConfiguration` method takes two arguments: old options, and new options (in both cases with severities removed). It should return an options object that (ideally) contains a merging of the two sets of options, in whatever way makes sense for the rule in question.
* When resolving configurations, ESLint core would load the rule, look for an `extendConfiguration` method, and call it with old and new options if it is present. If not, ESLint core will simply override with the new options, matching current behavior.

One possible challenge I am already seeing is that this requires CLIEngine to load and work with rules, instead of Linter. This may not be a problem if schema validation is also being handled by CLIEngine, but I can't remember if it is. I'm sure there are more issues with what I have proposed as well. But let's see if we can come up with something workable and get this sorted!

**Comment by not-an-aardvark** [comment]
Thanks for creating this issue, this is something I've wanted to discuss as well. Your proposal seems like a good start -- I have a few questions for discussion:

* Would it be possible for end users to override the rule from the shareable config? For example, if `no-restricted-syntax` had an `extendConfiguration` method like `(configA, configB) => configA.concat(configB)`, there would be no way for a user to get rid of the values from the old options list if the user didn't want those values to be part of the rule's config.
* How many distinct, useful `extendConfiguration` methods are there? It seems like the only one we've encountered is `(configA, configB) => configA.concat(configB)`.

For what it's worth, I had been planning to propose that we overload the severity slot:

```js
{
      "no-restricted-syntax": ["inherit", "FooStatement"]
}
```

...where the string `'inherit'` would be a special value that would cause the old config to be concatenated with the new config.

**Comment by platinumazure** [comment]
> * Would it be possible for end users to override the rule from the shareable config? For example, if `no-restricted-syntax` had an `extendConfiguration` method like `(configA, configB) => configA.concat(configB)`, there would be no way for a user to get rid of the values from the old options list if the user didn't want those values to be part of the rule's config.

It would be desirable to allow this for sure. I'm not really sure how yet. Maybe overloading the severity is the place to do so-- not sure.

> * How many distinct, useful `extendConfiguration` methods are there? It seems like the only one we've encountered is `(configA, configB) => configA.concat(configB)`.

Once or twice, I have longed for the ability to specify a partial object option, with the idea that it would be `Object.assign`'d with the extended rule's settings.

Example: I might wish to extend a `"max-len": ["error", { code: 80, ignoreTrailingComments: true }]` with an override to only set `ignoreTrailingComments: false` (regardless of the extended config's max-len code value).

> For what it's worth, I had been planning to propose that we overload the severity slot:

How would the user actually identify the severity he/she wants in that case?

**Comment by not-an-aardvark** [comment]
> How would the user actually identify the severity he/she wants in that case?

If we did overload the severity slot, I imagine the severity would always be the same as the original severity from the configuration being extended. The user could use the existing method of specifying a severity with full options, rather than using the new "inherit" behavior if they wanted to override the severity. Personally, I consider this to be mildly undesirable but not a complete blocker if we decided that this change was worthwhile.

Stepping back a bit, my design goals for an improvement here are:

1. When configuring rules that accept multiple independent options in an array (e.g. `no-restricted-syntax`), the user should be able to add additional options to an inherited configuration of the rule, without duplicating the inherited configuration in their own config file.

    * Configurability for other types of rules would also be nice, but I don't consider it a requirement.
    * Being able to change the severity of the rule at the same time as adding additional options would also be nice, but is not a requirement.
1. The change should be backwards-compatible with our existing config logic.
1. The user should still be able to completely override inherited options, rather than concatenating options, if they choose to do so.
1. The change should be reasonably simple for users to configure.
1. The change should not overly inhibit our ability to make future enhancements to config logic.

**Comment by Nokel81** [comment]
As for the changing severity how about the following:

1. Overriding both level and subrules:
```json
"no-restricted-syntax": ["error", "FooStatement"]
```

2. Inheriting with overriding level but extend sub-rules:
```json
"no-restricted-syntax": {
    "inherit": ["error", "FooStatement"]
}
```

3. Inheriting without overriding level and with extending sub-rules:
```json
"no-restricted-syntax": {
    "inherit": ["inherit", "FooStatement"]
}
```

4. Inheriting without overriding level and overriding sub-rules:
```json
"no-restricted-syntax": ["inherit", "FooStatement"]
```

**Comment by not-an-aardvark** [comment]
@Nokel81 I'm a bit averse to having two different forms for the configuration (object and array). The semantics of case 3 seem very non-obvious to me (the special word `"inherit"` is used in two different places with two different meanings).

Personally, I'm not sure it's necessary to cover all the possible combinations of overloads that a user might want -- in the worst-case scenario, they can always use the current behavior and overload everything.

**Comment by platinumazure** [comment]
@not-an-aardvark What about just prepending to the beginning of the array? `["inherit", "error", /* options */]`

In that case, the user has to specify the severity when overriding even if they want to keep it the same, but that is currently the case even without inherit/override support.

And I still think we could leave it to rules to implement an extend/override options method (which could, e.g., concat or `Object.assign` as we discussed earlier).

**Comment by ljharb** [comment]
It would be ideal if extending deep-merged all config args, including nested arrays and objects. One important use case is “no-restricted-*”, where it’s very valuable to be able to extend a preconfigured list of errors.

**Comment by platinumazure** [comment]
@ljharb Under my proposal, each rule could come up with its own implementation for extending configuration and could certainly support a deep merge if desired. I don't think it's always necessary to go that far, though-- some rules would be just fine with a shallow merge or concat. Also, who knows what plugin authors would want to do. So that's why I think it makes sense to let rules choose, based on what they know of their own schemas.

**Comment by not-an-aardvark** [comment]
@platinumazure That could work. Having two strings at the start seems a bit weird to me aesthetically (it would be odd if the severity were sometimes the first slot in the array and sometimes the second), but I realize there might be no ideal solution here.

Having said that, I'm still not convinced that it's a good idea to let rules choose their own "inherit" behavior. 

* A rule might not be aware of the best way to merge their configs. In fact, it seems like could be multiple ways (e.g. one for the union of reported errors between the two configs, and one for the intersection of reported errors).
* It seems like that this would be something a rule would have to implement and document, and if each rule behaved differently this could be confusing for users.

If we didn't need to support JSON configs, having a function value in a config would be a good solution so that the user would be able to provide the mapping themselves. However, I doubt there would be consensus on a feature only available to JS configs.

---

@ljharb I don't think that's workable because users still need to be able to override an extended config if they choose to do so. Also, it sometimes doesn't make sense to deep-merge two configs (e.g. config options can be mutually exclusive).

**Comment by ljharb** [comment]
Hmm :-/ requiring rules to opt in to being extensible seems unfortunate. Is there no sane default behavior?

**Comment by platinumazure** [comment]
> It seems like that this would be something a rule would have to implement and document, and if each rule behaved differently this could be confusing for users.

I see your point, but I am presently convinced that it is the only option that makes sense for the wide world of ESLint rules.

> Hmm :-/ requiring rules to opt in to being extensible seems unfortunate. Is there no sane default behavior?

I'm not sure there is, honestly, due to the size of the ecosystem and the expansive set of possible schemas out there. I'm worried that if we chose a particular "sane default", it would break a number of plugins.

Even in ESLint core alone, I think array concat and object assign are already two equally good extension methods, and we would easily use both. So again, I don't think we should just pick one.

**Comment by not-an-aardvark** [comment]
Could we choose both? Maybe we could have something like `"concat"` and `"merge"` strings.

**Comment by ljharb** [comment]
That’s what i meant by “deep merge” - use assign for objects and concat for arrays, recursively.

**Comment by platinumazure** [comment]
@not-an-aardvark I don't like that, because it implies that if a rule's schema expands and the rule needs to create a new extend behavior, `"concat"` and `"merge"` stop making sense semantically and the user would probably need to change their config. So in other words, we either have semantic nonsense in configs or introduce the possibility of breakage whenever a rule's schema changes.

I could get behind something that's more like an `"intersect"` vs `"union"` and rules could choose to support one, both, or neither, but that's basically my original proposal times two. Probably not a great idea.

@ljharb 

> That’s what i meant by “deep merge” - use assign for objects and concat for arrays, recursively.

Well, I'd be interested in seeing someone experiment with that in core rules to see how that would look. Maybe we could identify some early issues by trying that out, even if only conceptually.

That said, I'm still not convinced that rules won't need to specialize in many, many cases. For example, `indent` config is a 2-length array (I'm ignoring severity and possible extend/override keywords), first with the indentation figure, then object options. In *that* case, I would want to do item-by-item override or merge (override the indentation length, merge the object option).

**Comment by ljharb** [comment]
That’s what a deep merge would do, i think.

**Comment by not-an-aardvark** [comment]
A deep merge wouldn't work for rules like [`prefer-destructuring`](https://eslint.org/docs/rules/prefer-destructuring), which has two mutually-exclusive schema variants both represented by objects.

(Actually, I'm not sure it would work for rules like `no-restricted-syntax` either, unless we made some modifications to the typical `lodash.merge`-style algorithm. Two sets of options like `[{selector: 'foo'}]` and `[{selector: 'bar'}]` would just get merged to `[{selector: 'bar'}]`.)

**Comment by nzakas** [comment]
It looks like this conversation has stalled, so I'd like to see if anyone wants to keep it going?

If so, I'd suggest breaking this up into two different problems:

1. How can a rule like `no-restricted-syntax` tell ESLint that it wants options to be merged instead of overridden?
2. How can a user systematically pick and choose which pieces of an extended configuration to change?

I'd argue that both of these problems are worth investigating and solving one doesn't necessary solve the other. It seems like @platinumazure's original proposal solves the first problem, though I would call the method `mergeOptions` instead of `extendConfiguration` to make the intent clearer. It seems like this would solve the immediate-term case of `no-restricted-syntax` being very difficult to use with `extends` in any sort of rational way.

I'm not sure the second problem is actually solvable given the various forms that rule configuration can take, but I don't think that should block looking at the first problem.

Just my $0.02. We can also decide not to address either problem and just close the issue. :)

**Comment by not-an-aardvark** [comment]
> 1. How can a rule like no-restricted-syntax tell ESLint that it wants options to be merged instead of overridden?
> 1. How can a user systematically pick and choose which pieces of an extended configuration to change?

I don't think solving the first problem as stated would be a good idea, because it's important that users retain the ability to completely override an extended config rather than merging options if they choose to do so. So in either case, it seems like this would need to be a config change exposed to end users, rather than only being an addition to the rule API.

I think it might be better to reword the two problems as:

1. How can a user tell ESLint that it wants options for a rule like `no-restricted-syntax` to be merged instead of overridden, possibly with explicit support from the rule author?
1. How can a user systematically pick and choose which pieces of an extended configuration to change?

In this case, solving (2) would imply solving (1). I agree that (2) might not be worth solving/solvable. Personally I think (1) is worth solving (or at least I think it's worth evaluating potential solutions for).

---

What would you think about introducing special config keywords such as `"concat"` for various generic "merging" behaviors? The keyword could either replace the rule severity or precede it. For example, the following config would concatenate the inherited list of options for `foo-rule` with the array `['bar', 'baz']`:

```json
{
    "rules": {
        "foo-rule": ["concat", "error", "bar", "baz"]
    }
}
```

This would also give us leeway to add additional merging behaviors in the future as we identify them, by introducing new keywords.

As a downside, the above has the potential to decrease readability. As long as rules are configured with arrays, it seems like enhancing rule configurations will inevitably hurt readability. Another potential solution would be to bite the bullet and allow rules to be configured as objects, as suggested in https://github.com/eslint/eslint/issues/9192#issuecomment-370613244. (I was originally against this in https://github.com/eslint/eslint/issues/9192#issuecomment-370618606, but on reconsideration it seems like using objects would make it much easier to enhance rule configurations here, and also in the future if we want to make other enhancements.) That might look like this:

```json
{
    "rules": {
        "semi": {
            "severity": "error",
            "options": ["never"]
        },
        "foo-rule": {
            "inheritance": "concat",
            "options": ["bar", "baz"]
        }
    }
}
```

Of course, if we did start supporting objects we would need to continue to support arrays to avoid breaking all existing configs.

**Comment by ilyavolodin** [comment]
I think this is a problem worth solving, but I'm against solving it with adding a first (optional) argument. We state everywhere in our documentation that the first argument of the configuration array is always severity of the error. I think it would be very confusing if we were to break that promise. What if we just split configuration into separate object?

```js
{
  "rules": {
    "semi": ["never"],
    ....
    "inherit": {
      "no-restricted-syntax": ["error", "ForOfStatement"],
      ...
    }
}
```

This will allow for much clearer backwards compatibility. Inherit could also be a top level node if you don't like nesting it under `rules`, although I would call it `inheritedRules` in that case.

**Comment by nzakas** [comment]
@not-an-aardvark I think the first problem is worth looking at. What it gets at is the *default* way that a rule behaves when overridden. The default right now is just to squash all options. Solving the first problem would immediately improve the user experience for `no-restricted-syntax` while giving us more time to solve the second problem.

In regards to the second problem, I'm with @ilyavolodin in that I don't want to prepend another option to rule configuration. That seems very confusing to me. I'd also prefer not to have something like `inherit` in the place where a rule name has always been. Both of those solutions break explicit contracts we've had with our users about what to expect where in configs.

That said, I think this whole conversation has focused too much on `no-restricted-syntax`. We have a ton of rules, all with very different configuration option formats: some with strings, some with objects, some with arrays. I'm skeptical at this point that there is a solution to the second problem that will work correctly for all of the various ways rules can be configured.  For instance, how would you expect to inherit just some of the options from `indent`?

Before going any further, what I'd like to do is get a consensus on whether we are going to try to solve both problems I mentioned or only one so we can focus the conversation in a particular direction.

**Comment by not-an-aardvark** [comment]
I think we're in agreement that it's not worth pursuing a solution to the second problem.

To clarify my concern with the first problem from https://github.com/eslint/eslint/issues/9192#issuecomment-439583495: It's not clear to me how this would just be a default, because it would apply whenever both a user and a shareable config configure the same rule.

For example, if a shareable config contains:

```
{
    "rules": {
        "no-restricted-syntax": ["error", "foo", "bar", "baz"]
        ...other rules
    }
}
```

...and the user is extending the shareable config, and wants their final resolved configuration to look like this:

```
{
    "rules": {
        "no-restricted-syntax": ["error", "foo", "bar"]
        ...other rules
    }
}
```

Currently, the user can obtain their desired config by simply declaring it in their config file, i.e. by adding `"no-restricted-syntax": ["error", "foo", "bar"]`. 

If we allowed `no-restricted-syntax` to express a behavior like "options should be concatenated when overriding", then adding `"no-restricted-syntax": ["error", "foo", "bar"]` to the end user's config would result in a final resolved configuration of `["foo", "bar", "baz", "foo", "bar"]`. In other words, it would be impossible for the end user to express their desired configuration for a rule because the user's shareable config already specified a different configuration for it. This would be very undesirable, in my opinion -- shareable configs currently allow the user to override any extended rule configuration, and it's very common for projects to take advantage of that for project-specific deviations from a general styleguide.

**Comment by nzakas** [comment]
I see, so what you're saying is that right now people can effective "reset" the configuration for a rule today, and that ability would be removed if we solve the first problem (aka merging of options by the rule itself). That makes sense.

So after all of this, it sounds like maybe we don't want to solve either of the two problems at this point?

**Comment by ljharb** [comment]
It’s still a big pain point to require resetting of entire rules in order to tweak something from a shared config.

Perhaps we could help both problems by providing a) a way for a rule to *opt in* to being partially overrideable - like, perhaps, a function that accepts partial options - and b) a schema for a config to specify, independently of its normal config, partial options that would get passed into that function?

That way it would both not conflict with the existing “override” approach, and would be safe since each rule could determine for itself if it even accepted partial options, or if so, how to merge them.

**Comment by not-an-aardvark** [comment]
My suggestion would be to attempt to solve a modified version of the first problem instead:

> How can a user tell ESLint that it wants options for a rule like no-restricted-syntax to be merged instead of overridden, possibly with explicit support from the rule author?

In other words, this would need to be an end-user-facing change, not just an enhancement to the rule API. That's why I was proposing enhancements to the config format in https://github.com/eslint/eslint/issues/9192#issuecomment-439596455.

**Comment by nzakas** [comment]
Okay. What I'd suggest at this point is that we close this issue and open a new one that is laser-focused on that problem so we don't get stuck on the meandering middle of this thread.

I kind of get what you're suggesting in that comment, but it seems too high level for any real evaluation at this point. I'd personally want to see a proposal that would work for a selection of 3-5 different rules with different option formats.

(Of course another option is to go straight to an RFC, which would be necessary anyway based on adding a new features to the core.)
